Replace is room available

mebays opened the pull request:
Noticed this was deprecated I think I got the logic down to something that would resemble the current functionality.

mebays committed:
Replace isRoomAvailable

Deprecated function call.

mebays committed:
Run linter

mebays committed:
Move function to room meta

duckymirror commented:
Maybe I'm just blind but I can't find the method `Room.isAvailable()` in the docs. Did you mean this?
![image](https://user-images.githubusercontent.com/33291860/87174736-ebea9d00-c2d7-11ea-89aa-1e2865a75daf.png)

duckymirror commented on src/extends/room/meta.js:
```diff
@@ -182,3 +182,11 @@ Room.isCuldesac = function (roomName, entrance) {
+  if (Game.gcl.level > GCL_NOVICE) {
+    return Game.map.getRoomStatus(name).status === 'normal'
+  } else {
+    return Game.map.getRoomStatus(name).status === 'novice'
```
Wouldn't this prevent scouting in novice areas once the user is above GCL3?

duckymirror commented on src/extends/room/meta.js:
```diff
@@ -182,3 +182,11 @@ Room.isCuldesac = function (roomName, entrance) {
+
+Room.isAvailable = function (name) {
+  if (Game.gcl.level > GCL_NOVICE) {
+    return Game.map.getRoomStatus(name).status === 'normal'
```
What about respawn areas?

mebays replied:
If `isClaimable` gets thje GCL_NOVICE check would it be better to use simpler logic and just return `Game.map.getRoomStatus(name).status !== 'closed'`

duckymirror replied:
Yes, I think so. In the future we could add a method that checks whether a room is accessible based on whether it's in an area Quorum isn't in.

duckymirror commented on src/extends/room/meta.js:
```diff
@@ -182,3 +182,11 @@ Room.isCuldesac = function (roomName, entrance) {
 }
+
+Room.isAvailable = function (name) {
+  if (Game.gcl.level > GCL_NOVICE) {
```
Did you mean `>=`? You can't claim a new room in a novice area when your GCL === GCL_NOVICE.

duckymirror commented on src/extends/room/meta.js:
```diff
@@ -164,7 +164,7 @@ Room.isHallway = function (name) {
 
 Room.isClaimable = function (name) {
-  if (!Game.map.isRoomAvailable(name)) {
+  if (!Room.isAvailable(name)) {
```
I'd move the GCL_NOVICE check here.

mebays replied:
I can move the GCL_NOVICE check here.

duckymirror reviewed (changes requested):
Risking that I'm missing something again, this is what I see.

mebays committed:
Update functions

- isClaimable now considers the GCL level
- isAvailable only checks if the room is not closed in order to consider
  all possibilities (normal, novice, respawn)

duckymirror commented on src/extends/room/meta.js:
```diff
@@ -164,6 +164,12 @@ Room.isHallway = function (name) {
 }
 
 Room.isClaimable = function (name) {
+  if (Game.gcl.level >= GCL_NOVICE && Game.map.getRoomStatus(name).status !== 'normal') {
```
I didn't see that either yesterday: We should check for owned rooms instead of current GCL.

duckymirror commented on src/extends/room/meta.js:
```diff
@@ -164,6 +164,12 @@ Room.isHallway = function (name) {
+  if (Game.gcl.level >= GCL_NOVICE && Game.map.getRoomStatus(name).status !== 'normal') {
+    return false
+  }
+  if (Game.gcl.level < GCL_NOVICE && Game.map.getRoomStatus(name).status !== 'novice') {
```
Did you mean the opposite? If we have >= GCL_NOVICE rooms and the room is in a novice area?

duckymirror commented on src/extends/room/meta.js:
```diff
@@ -164,6 +164,12 @@ Room.isHallway = function (name) {
 }
 
 Room.isClaimable = function (name) {
+  if (Game.gcl.level >= GCL_NOVICE && Game.map.getRoomStatus(name).status !== 'normal') {
```
What is this condition for?

mebays replied:
This will be updated to >= GCL_NOVICE && Game.map.getRoomStatus(name).status === 'novice' in order to check if the room can be claimed. Like the above comment suggested.

mebays committed:
Update logic

duckymirror commented on src/extends/room/meta.js:
```diff
@@ -164,7 +164,13 @@ Room.isHallway = function (name) {
+  if (Game.gcl.level >= GCL_NOVICE && Game.map.getRoomStatus(name).status === 'novice') {
+    return false
+  }
+  if (Game.gcl.level < GCL_NOVICE && Game.map.getRoomStatus(name).status === 'normal') {
```
This condition will prevent Quorum from expanding outside novice and respawn areas if its GCL is less than GCL_NOVICE.

duckymirror commented on src/extends/room/meta.js:
```diff
@@ -164,7 +164,13 @@ Room.isHallway = function (name) {
 
 Room.isClaimable = function (name) {
-  if (!Game.map.isRoomAvailable(name)) {
+  if (Game.gcl.level >= GCL_NOVICE && Game.map.getRoomStatus(name).status === 'novice') {
```
This condition will prevent Quorum from claiming any rooms in novice areas if its GCL is GCL_NOVICE or higher. This means that it can't claim a third room (or even a second one if it spawned in the novice area with GCL 3).
